pkg/controllers/state/state.go

```diff
@@ -0,0 +1,241 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"knative.dev/pkg/logging"
+
+	"golang.org/x/time/rate"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	podutils "github.com/aws/karpenter/pkg/utils/pod"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// Cluster maintains cluster state that is often needed but expensive to compute.
+type Cluster struct {
+	ctx         context.Context
+	rateLimiter *rate.Limiter
+	kubeClient  client.Client
+
+	// Pod Specific Tracking
+	knownPods        sync.Map // mapping of pod namespaced name to struct{} used to track if we've seen a pod before
+	antiAffinityPods sync.Map // mapping of pod namespaced name to *v1.Pod of pods that have required anti affinities
+
+	// Node Status & Pod -> Node Binding
+	mu       sync.RWMutex
+	nodes    map[string]*Node                // node name -> node
+	bindings map[types.NamespacedName]string // pod namespaced named -> pod
+}
+
+func NewCluster(ctx context.Context, client client.Client) *Cluster {
+	s := &Cluster{
+		ctx:         ctx,
+		kubeClient:  client,
+		rateLimiter: rate.NewLimiter(rate.Every(10*time.Second), 1),
+		nodes:       map[string]*Node{},
+		bindings:    map[types.NamespacedName]string{},
+	}
+	return s
+}
+
+// Node is a cached version of a node in the cluster that maintains state which is expensive to compute every time it's
+// needed.  This currently contains node utilization across all of the allocatable resources, but will soon be used to
+// compute topology information.
+type Node struct {
+	Node        *v1.Node
+	requested   v1.ResourceList
+	allocatable v1.ResourceList
+
+	podRequests map[types.NamespacedName]v1.ResourceList
+}
+
+// Requested returns the total amount of resources requested by pods that have been bound to the node.
+func (n Node) Requested() v1.ResourceList {
+	return n.requested
+}
+
+// ForPodsWithAntiAffinity calls the supplied function once for each pod with required anti affinity terms that is
+// currently bound to a node. The pod returned may not be up to date with respect to status, however since the
+// anti-affinity terms can't be modified, they will be correct.
+func (c *Cluster) ForPodsWithAntiAffinity(fn func(p *v1.Pod, n *v1.Node) bool) {
+	c.antiAffinityPods.Range(func(key, value interface{}) bool {
+		pod := value.(*v1.Pod)
+		c.mu.RLock()
+		defer c.mu.RUnlock()
+		nodeName, ok := c.bindings[client.ObjectKeyFromObject(pod)]
+		if !ok {
+			return true
+		}
+		node, ok := c.nodes[nodeName]
+		if !ok {
+			// if we receive the node deletion event before the pod deletion event, this can happen
```

Comment:
Can't we delete all pods on the node on the node deletion event?

Reply:
We could, I wrote it that way the first time but then deleted it because it was more code and it would take care of itself in a short time anyway.  Happy to do it either way.